### ISSUE TITLE
Make no-restricted-imports apply to dynamic imports

### ISSUE THREAD

**Issue by robatwilliams** [eslint/eslint#13594]
**What rule do you want to change?**
no-restricted-imports

**Does this change cause the rule to produce more or fewer warnings?**
More

**How will the change be implemented? (New option, new default behavior, etc.)?**
New default behaviour would make sense, although could be an opt-in via option too.

**Please provide some example code that this change will affect:**

```js
await import('restricted');
```

**What does the rule currently do for this code?**
Nothing.

**What will the rule do after it's changed?**
Report an error just as if it was a sync import.

**Are you willing to submit a pull request to implement this change?**
Maybe

**Comment by robatwilliams** [comment]
> May be behind an option with something like dynamicImport: Boolean makes sense as the dynamic import is available from 2020 ECMA version.
Or perhaps enable it by default if version >= 2020.

>  I'm not sure how useful this feature would be
I'll share our use case... We're doing webpack code splitting using dynamic imports, and we use no-restricted-imports with patterns to enforce that things aren't imported from places they shouldn't be - which would result in them being in the wrong (or multiple) bundles.

**Comment by mdjermanovic** [comment]
> > May be behind an option with something like dynamicImport: Boolean makes sense as the dynamic import is available from 2020 ECMA version.
> 
> Or perhaps enable it by default if version >= 2020.

New default behavior would be a breaking change, because the rule currently doesn't check dynamic imports and ESLint already supports dynamic imports for over a year now. 

Since this change results in ESLint reporting more linting errors, and it isn't a bug fix, per our [semver-policy](https://github.com/eslint/eslint#semantic-versioning-policy) we would have to either wait for the next major version, or add an option with a default value that retains the current behavior of this rule.

`ecmaVersion >= 2020` is basically the same as not checking `ecmaVersion` at all, because this syntax isn't available in `ecmaVersion < 2020`.

> > I'm not sure how useful this feature would be
> 
> I'll share our use case... We're doing webpack code splitting using dynamic imports, and we use no-restricted-imports with patterns to enforce that things aren't imported from places they shouldn't be - which would result in them being in the wrong (or multiple) bundles.

Sorry, I wasn't clear enough! There are certainly valid use cases, like with static imports.

The problem is that, unlike with static imports, we often can't know the import source. Dynamic imports have dynamic specifiers; `import()` syntax allows expressions inside `()`:

We could say that the new option applies only to dynamic imports with static specifiers (i.e. string literals, like `import("foo")`).

Still, it might be confusing when combined with other configuration options. For example, user might expect an error on the dynamic import in the following case (assuming that the new option is enabled):

```js
/*eslint no-restricted-imports: ["error", { "patterns": ["foo/*"] }]*/

import a from "foo/a"; // error

import(`foo/${bar}`); // error?
```

Also, the rule won't be able to restrict just certain names from a module:

```js
/*eslint no-restricted-imports: ["error", { "name": "foo", importNames: ["a"] }]*/

import { a } from "foo"; // error

import { b } from "foo"; // ok

import * as foo_ns from "foo"; // error, because `foo_ns` provides access to `a`

import("foo"); // always error?
```

**Comment by anikethsaha** [comment]
> Makes sense, but I'm not sure how useful this feature would be.
> 
> In particular, it seems it would be able to check only statically known import source strings. For example, it would probably allow imports such as `` import(`${foo}/bar`) ``, or just `import(pathToModule)`. Though, we could use [`eslint-utils.getStringIfConstant`](https://eslint-utils.mysticatea.dev/api/ast-utils.html#getstringifconstant) to check more than just string literals, but it would be still limited.
> 
> Also, any configuration that aims to restrict only certain names from a module (`"importNames"` option) would probably restrict the whole module from being dynamically imported, similar to how specifying `"importNames"` already disallows `import *` from that module.

I am +1 for using `getstringifconstant` with the scope argument. Yea there would be some known limitation to it and I think we can document that as well. 

> ```js
> /*eslint no-restricted-imports: ["error", { "patterns": ["foo/*"] }]*/
> 
> import a from "foo/a"; // error
> 
> import(`foo/${bar}`); // error?
> ```

How about applying the `patterns` on the output of `getstringifconstant` ? so I guess it wont check `import(`foo/${bar}`);`.



> Also, the rule won't be able to restrict just certain names from a module:
> 
> ```js
> /*eslint no-restricted-imports: ["error", { "name": "foo", importNames: ["a"] }]*/
> 
> import { a } from "foo"; // error
> 
> import { b } from "foo"; // ok
> 
> import * as foo_ns from "foo"; // error, because `foo_ns` provides access to `a`
> 
> import("foo"); // always error?
> ```

IMO, not to check ` import("foo");` either when `importNames` is mentioned ?


That being said, I can see there seems good number of cases where it will not work. But in others, it will be helpful. 
If we are adding support for this, it would be better to put behind an option.

**Comment by mdjermanovic** [comment]
> ```js
> /*eslint no-restricted-imports: ["error", { "patterns": ["foo/*"] }]*/
> 
> import a from "foo/a"; // error
> 
> import(`foo/${bar}`); // error?
> ```
>
> How about applying the patterns on the output of getstringifconstant ? so I guess it wont check import(foo/${bar});

It wouldn't report ``import(`foo/${bar}`)``, but I have concerns that users might expect otherwise, as it matches `foo/*`.

I guess we could document this in Known Limitations.

>> ```js
>> /*eslint no-restricted-imports: ["error", { "name": "foo", importNames: ["a"] }]*/
>> 
>> import { a } from "foo"; // error
>> 
>> import { b } from "foo"; // ok
>> 
>> import * as foo_ns from "foo"; // error, because `foo_ns` provides access to `a`
>> 
>> import("foo"); // always error?
>> ```
>
> IMO, not to check import("foo"); either when importNames is mentioned ?
> Not checking would be useful in order to avoid false results

If it doesn't disallow `import("foo")`, then it's basically ignoring user's configuration `{ "name": "foo", importNames: ["a"] }`.

If it does disallow `import("foo")`, then there's no way to get `b` through dynamic imports, although the configuration intends to disallow only `a`. 

Either way, it seems that the new option would be incompatible with an existing option (`"importNames"`).

**Comment by robatwilliams** [comment]
Limiting to static specifiers as a known limitation sounds good, there must be a few rules that have such a limitation (e.g. `no-dupe-keys`). There are all kinds of expressions that could go inside an `import( )` so supporting some of them would likely be more surprising than not supporting them at all. And checking the `patterns` against that static value also sounds good.

The choice between allowing/disallowing when `importNames` is specified I think depends on the importance of soundness vs. false positives when ESLint can't definitively decide. I think avoiding false positives is more the "ESLint way"?

Perhaps something like "The `dynamicImports` option makes the rule apply to `import()` expressions where the source is statically specified. Restrictions that include `importNames` are not applied."

**Comment by mdjermanovic** [comment]
> Limiting to static specifiers as a known limitation sounds good, there must be a few rules that have such a limitation (e.g. `no-dupe-keys`). There are all kinds of expressions that could go inside an `import( )` so supporting some of them would likely be more surprising than not supporting them at all. And checking the `patterns` against that static value also sounds good.
>
>
>
> The choice between allowing/disallowing when importNames is specified I think depends on the importance of soundness vs. false positives when ESLint can't definitively decide. I think avoiding false positives is more the "ESLint way"?

I think both are important. A rule with a lot of false positives would be unusable. On the other hand, a rule with a lot of false negatives is unreliable, and if it can't catch the common cases then it might be better not to implement the rule (or an option) than to set false expectations. Dynamic specifiers are listed as one of the common use cases in the [proposal](https://github.com/tc39/proposal-dynamic-import) (as "factors only known at runtime").

`importNames` also seem really problematic for dynamic imports. Specifying `importNames` looks like an intention to use something else from that module, so completely disallowing that module probably isn't desirable. But, allowing everything would be inconsistent with the actual behavior for `import *`, and would be basically ignoring the user's configuration.

I don't have a strong opinion about the new option. It could be useful, but still limited and incompatible with other options. It's interesting that we didn't even consider updating this rule when dynamic imports were added to the core (#11803).

**Comment by wellrus** [comment]
I'd love to be able to validate dynamic imports.

I will give the following example as an example. The application is written in vue. The application is split into parts and loaded dynamically. All this takes place in a mono-repository. Moreover, the packages inside the monorepository have a logical connection with each other. Suppose you cannot import dependencies from the shell package into the app package.

Direct imports can be validated, but dynamic imports cannot = (

monorepo/packages/app/App.vue
`
<template>
  <Shell>
    <router-view/>
  </Shell>
</template>

<script lang="ts">
import { defineComponent, defineAsyncComponent } from 'vue';
import DirectShellImport from '@monorepo/shell'; // ESLint: '@monorepo/shell' import is restricted from being used.(no-restricted-imports)

export default defineComponent({
  name: 'App',
  components: {
    DirectShellImport: DirectShellImport,
    Shell: defineAsyncComponent(() => import('@monorepo/shell'))  // dynamic import still valid
  },
  setup () {
    console.log('App call setup method');
  }
});
</script>

<style scoped lang="scss">
</style>
`

eslint-rule
`
            rules: {
                "no-restricted-imports": ["error", {
                    "paths": ["@monorepo/shell"],
                    "patterns": ["@monorepo/shell"]
                }]
            }
`

**Comment by wellrus** [comment]
Obviously, it is difficult to validate dynamic imports if the path is calculated. And validation is not possible for all cases.

// difficult but possible
const package = 'shell';
import('@monorepo/'+package ))

// impossible
const package = process.env.IMPORT_NAME;
import('@monorepo/'+package ))

However, it is absolutely possible to validate imports where dependencies are explicitly specified.

import('@monorepo/shell'))

**Comment by mdjermanovic** [comment]
If this enhancement doesn't reach [consensus](https://eslint.org/docs/maintainer-guide/issues#accepting-issues), an alternative is to configure [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax) rule.

For example, this disallows importing `foo` and anything from `bar/`:

```js
/* eslint "no-restricted-syntax": [ "error",
  {
    "selector": "ImportExpression > Literal.source[value='foo']",
    "message": "Don't import 'foo'."
  },
  {
    "selector": "ImportExpression > Literal.source[value=/^bar\\x2F/]",
    "message": "Don't import 'bar/*'."
  }
]*/

import("foo"); // Don't import 'foo'. (no-restricted-syntax)

import("bar/baz"); // Don't import 'bar/*'. (no-restricted-syntax)
```

[Online demo link](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tcmVzdHJpY3RlZC1zeW50YXhcIjogWyBcImVycm9yXCIsXG4gIHtcbiAgICBcInNlbGVjdG9yXCI6IFwiSW1wb3J0RXhwcmVzc2lvbiA+IExpdGVyYWwuc291cmNlW3ZhbHVlPSdmb28nXVwiLFxuICAgIFwibWVzc2FnZVwiOiBcIkRvbid0IGltcG9ydCAnZm9vJy5cIlxuICB9LFxuICB7XG4gICAgXCJzZWxlY3RvclwiOiBcIkltcG9ydEV4cHJlc3Npb24gPiBMaXRlcmFsLnNvdXJjZVt2YWx1ZT0vXmJhclxcXFx4MkYvXVwiLFxuICAgIFwibWVzc2FnZVwiOiBcIkRvbid0IGltcG9ydCAnYmFyLyonLlwiXG4gIH1cbl0qL1xuXG5pbXBvcnQoXCJmb29cIik7IC8vIERvbid0IGltcG9ydCAnZm9vJy4gKG5vLXJlc3RyaWN0ZWQtc3ludGF4KVxuXG5pbXBvcnQoXCJiYXIvYmF6XCIpOyAvLyBEb24ndCBpbXBvcnQgJ2Jhci8qJy4gKG5vLXJlc3RyaWN0ZWQtc3ludGF4KVxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUtZWxzZS1pZiI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW1wb3J0LWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1pc2xlYWRpbmctY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXByb3RvdHlwZS1idWlsdGlucyI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zZXR0ZXItcmV0dXJuIjoyLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVudXNlZC1sYWJlbHMiOjIsIm5vLXVudXNlZC12YXJzIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnt9fX0=)

**Comment by robatwilliams** [comment]
Would it at least be worth documenting that the current rule is for static imports only? Or perhaps the lack of dynamic imports in the examples is enough.

**Comment by mdjermanovic** [comment]
> Would it at least be worth documenting that the current rule is for static imports only?

Yes, I think it's a great idea. PR is welcome!
